File: library/src/main/java/me/shkschneider/skeleton/java/JobManager.java
package me.shkschneider.skeleton.java;

import android.annotation.TargetApi;
import android.app.job.JobInfo;
import android.app.job.JobScheduler;
import android.app.job.JobService;
import android.content.ComponentName;
import android.os.PersistableBundle;
import android.support.annotation.NonNull;

import me.shkschneider.skeleton.helper.AndroidHelper;
import me.shkschneider.skeleton.helper.ContextHelper;
import me.shkschneider.skeleton.helper.SystemServices;

/**
 * <service
 *  android:name=".MyJobService"
 *  android:permission="android.permission.BIND_JOB_SERVICE" />
 */
@TargetApi(AndroidHelper.API_21)
public class JobManager {

    public static void cancelAll() {
        SystemServices.jobScheduler().cancelAll();
    }

    public static void cancel(final int id) {
        SystemServices.jobScheduler().cancel(id);
    }

    private int mId;
    private JobInfo.Builder mBuilder;

    public JobManager(final int id, @NonNull final Class<? extends JobService> cls) {
        mId = id;
        final ComponentName componentName = new ComponentName(ContextHelper.applicationContext(), cls.getClass());
        mBuilder = new JobInfo.Builder(id, componentName);
    }

    @SkHide
    public JobManager setExtras(final PersistableBundle extras) {
        mBuilder.setExtras(extras);
        return this;
    }

    @SkHide
    public JobManager setRequiredNetworkType(final int networkType) {
        mBuilder.setRequiredNetworkType(networkType);
        return this;
    }

    @SkHide
    public JobManager setRequiresCharging(final boolean requiresCharging) {
        mBuilder.setRequiresCharging(requiresCharging);
        return this;
    }

    @SkHide
    public JobManager setRequiresDeviceIdle(final boolean requiresDeviceIdle) {
        mBuilder.setRequiresDeviceIdle(requiresDeviceIdle);
        return this;
    }

    @SkHide
    public JobManager addTriggerContentUri(@NonNull final JobInfo.TriggerContentUri uri) {
        mBuilder.addTriggerContentUri(uri);
        return this;
    }

    @SkHide
    public JobManager setTriggerContentUpdateDelay(final long durationMs) {
        mBuilder.setTriggerContentUpdateDelay(durationMs);
        return this;
    }

    @SkHide
    public JobManager setTriggerContentMaxDelay(final long durationMs) {
        mBuilder.setTriggerContentMaxDelay(durationMs);
        return this;
    }

    @SkHide
    public JobManager setPeriodic(final long intervalMillis) {
        mBuilder.setPeriodic(intervalMillis);
        return this;
    }

    @SkHide
    public JobManager setPeriodic(final long intervalMillis, final long flexMillis) {
        mBuilder.setPeriodic(intervalMillis, flexMillis);
        return this;
    }

    @SkHide
    public JobManager setMinimumLatency(final long minLatencyMillis) {
        mBuilder.setMinimumLatency(minLatencyMillis);
        return this;
    }

    @SkHide
    public JobManager setOverrideDeadline(final long maxExecutionDelayMillis) {
        mBuilder.setOverrideDeadline(maxExecutionDelayMillis);
        return this;
    }

    @SkHide
    public JobManager setBackoffCriteria(final long initialBackoffMillis, final int backoffPolicy) {
        mBuilder.setBackoffCriteria(initialBackoffMillis, backoffPolicy);
        return this;
    }

    @SkHide
    public JobManager setPersisted(final boolean isPersisted) {
        mBuilder.setPersisted(isPersisted);
        return this;
    }

    public boolean schedule() {
        final JobInfo jobInfo = mBuilder.build();
        mId = jobInfo.getId();
        return (SystemServices.jobScheduler().schedule(jobInfo) == JobScheduler.RESULT_SUCCESS);
    }

    public void cancel() {
        SystemServices.jobScheduler().cancel(mId);
    }

}

File: library/src/main/kotlin/me/shkschneider/skeleton/java/JobManager.kt
package me.shkschneider.skeleton.java

import android.Manifest
import android.annotation.TargetApi
import android.app.job.JobInfo
import android.app.job.JobScheduler
import android.app.job.JobService
import android.content.ComponentName
import android.os.PersistableBundle
import android.support.annotation.RequiresPermission

import me.shkschneider.skeleton.helper.AndroidHelper
import me.shkschneider.skeleton.helper.ContextHelper
import me.shkschneider.skeleton.helper.SystemServices

/**
 * <service android:name=".MyJobService" android:permission="android.permission.BIND_JOB_SERVICE"></service>
 */
@TargetApi(AndroidHelper.API_21)
class JobManager {

    private var id: Int
    private val builder: JobInfo.Builder

    constructor(id: Int, cls: Class<out JobService>) {
        this.id = id
        builder = JobInfo.Builder(this.id, ComponentName(ContextHelper.applicationContext(), cls::class.java))
    }

    @SkHide
    fun setExtras(extras: PersistableBundle): JobManager {
        builder.setExtras(extras)
        return this
    }

    @SkHide
    fun setRequiredNetworkType(networkType: Int): JobManager {
        builder.setRequiredNetworkType(networkType)
        return this
    }

    @SkHide
    fun setRequiresCharging(requiresCharging: Boolean): JobManager {
        builder.setRequiresCharging(requiresCharging)
        return this
    }

    @SkHide
    fun setRequiresDeviceIdle(requiresDeviceIdle: Boolean): JobManager {
        builder.setRequiresDeviceIdle(requiresDeviceIdle)
        return this
    }

    @TargetApi(AndroidHelper.API_24)
    @SkHide
    fun addTriggerContentUri(uri: JobInfo.TriggerContentUri): JobManager {
        builder.addTriggerContentUri(uri)
        return this
    }

    @TargetApi(AndroidHelper.API_24)
    @SkHide
    fun setTriggerContentUpdateDelay(durationMs: Long): JobManager {
        builder.setTriggerContentUpdateDelay(durationMs)
        return this
    }

    @TargetApi(AndroidHelper.API_24)
    @SkHide
    fun setTriggerContentMaxDelay(durationMs: Long): JobManager {
        builder.setTriggerContentMaxDelay(durationMs)
        return this
    }

    @SkHide
    fun setPeriodic(intervalMillis: Long): JobManager {
        builder.setPeriodic(intervalMillis)
        return this
    }

    @TargetApi(AndroidHelper.API_24)
    @SkHide
    fun setPeriodic(intervalMillis: Long, flexMillis: Long): JobManager {
        builder.setPeriodic(intervalMillis, flexMillis)
        return this
    }

    @SkHide
    fun setMinimumLatency(minLatencyMillis: Long): JobManager {
        builder.setMinimumLatency(minLatencyMillis)
        return this
    }

    @SkHide
    fun setOverrideDeadline(maxExecutionDelayMillis: Long): JobManager {
        builder.setOverrideDeadline(maxExecutionDelayMillis)
        return this
    }

    @SkHide
    fun setBackoffCriteria(initialBackoffMillis: Long, backoffPolicy: Int): JobManager {
        builder.setBackoffCriteria(initialBackoffMillis, backoffPolicy)
        return this
    }

    @RequiresPermission(Manifest.permission.RECEIVE_BOOT_COMPLETED)
    @SkHide
    fun setPersisted(isPersisted: Boolean): JobManager {
        builder.setPersisted(isPersisted)
        return this
    }

    fun schedule(): Boolean {
        val jobInfo = builder.build()
        id = jobInfo.id
        return SystemServices.jobScheduler()?.schedule(jobInfo) == JobScheduler.RESULT_SUCCESS
    }

    fun cancel() {
        SystemServices.jobScheduler()?.cancel(id)
    }

    companion object {

        fun cancelAll() {
            SystemServices.jobScheduler()?.cancelAll()
        }

        fun cancel(id: Int) {
            SystemServices.jobScheduler()?.cancel(id)
        }

    }

}
